syntax = "proto3";

package backend.api;
import "google/protobuf/empty.proto";
import "google/protobuf/timestamp.proto";


service TilingAPI {
    rpc GetTiles(google.protobuf.Empty) returns (ListMetaTile);
    rpc AddTile(PlaylistName) returns (MetaTile);
    rpc DelTile(MetaTile) returns (google.protobuf.Empty);
    rpc UpdTile(MetaTile) returns (MetaTile);
}

service MpdAPI {
    rpc Connect(ConnectCredentials) returns (ConnectionDetails);
    rpc Connected(google.protobuf.Empty) returns (ConnectionDetails);
    rpc GetActiveGroup(google.protobuf.Empty) returns (PlaylistGroup);
    rpc GetPlaylists(PlaylistGroup) returns (ListPlaylistName);
}

message ConnectCredentials {
    string socket = 1;
    string password = 2;
}

message ConnectionDetails {
    ConnectionStatus status = 1;
}

enum ConnectionStatus {
    Connected = 0;
    FailedToConnect = 1;
    NoMpdBinary = 2;
}

message UUID {
    bytes value = 1;
}

message ListMetaTile {
    repeated MetaTile metatile = 1;
}

message PlaylistName {
    string name = 1;
}

message ListPlaylistName {
    repeated PlaylistName playlist = 1;
}

message PlaylistGroup {
    string name = 1;
}

message Playlist {
    repeated Song song = 1;
}

message Song {
    int64 id = 1;
    string directory = 2;
    string file = 3;
    int64 time = 4;
    float duration = 5;
    google.protobuf.Timestamp lastmodified = 6;
    string format = 7;
    string artist = 8;
    string albumartist = 9;
    string title = 10;
    string album = 11;
    int64 track = 12;
    string date = 13;
    string genre = 14;
    string composer = 15;
    int64 disc = 16;
}

message MetaTile {
    string name = 1;
    bool locked = 2;
    string order_by_col = 3;
    string order_by_asc = 4; 
}